### PR TITLE
Optimize series response format by using repeated fileds.

### PR DESCRIPTION
**Pull request by jeschkies**
**What this PR does / why we need it**:
The Protobuf map type is encodied as a repeated field of map entries. Decoding them to a slice is much faster than decoding them into a map. Since Loki is not using the fast key check for a map we can use the slice decoding.

This change also allows us to decode the JSON directly into the right protobuf struct. This doulbes the JSON decoding speed and reduces the memory pressure by ~40%.

```
› go test -bench=. -run=^$ -count=10 ./pkg/querier/queryrange > before.log
› go test -bench=. -run=^$ -count=10 ./pkg/querier/queryrange > after.log
› benchstat before.log after.log
goos: darwin
goarch: arm64
pkg: github.com/grafana/loki/pkg/querier/queryrange
                                                             │  before.log  │              after.log               │
                                                             │    sec/op    │    sec/op     vs base                │
ResponseMerge/mergeStreams_unlimited-10                         32.36m ± 0%   32.63m ±  2%        ~ (p=0.393 n=10)
ResponseMerge/mergeOrderedNonOverlappingStreams_unlimited-10    1.050m ± 1%   1.080m ±  3%   +2.84% (p=0.005 n=10)
ResponseMerge/mergeStreams_limited-10                           33.02m ± 0%   32.60m ±  1%   -1.29% (p=0.004 n=10)
ResponseMerge/mergeOrderedNonOverlappingStreams_limited-10      15.11m ± 0%   15.07m ±  0%        ~ (p=0.075 n=10)
_CodecDecodeLogs-10                                             4.395m ± 1%   4.364m ±  0%   -0.72% (p=0.005 n=10)
_CodecDecodeSamples-10                                          16.97m ± 0%   16.84m ±  2%   -0.77% (p=0.023 n=10)
_CodecDecodeSeries/application/vnd.google.protobuf-10           745.8µ ± 8%   736.8µ ± 12%        ~ (p=0.739 n=10)
_CodecDecodeSeries/application/json;_charset=utf-8-10           15.37m ± 1%   10.60m ±  0%  -31.03% (p=0.000 n=10)
_MergeResponses-10                                             1186.9m ± 2%   149.8m ±  1%  -87.38% (p=0.000 n=10)
_UnwrapSeries-10                                                9.399m ± 1%   4.049m ±  0%  -56.92% (p=0.000 n=10)
_DecodeMergeEncodeCycle-10                                      666.0m ± 3%   194.6m ±  0%  -70.79% (p=0.000 n=10)
geomean                                                         18.87m        12.51m        -33.70%

                                                      │  before.log   │              after.log               │
                                                      │     B/op      │     B/op      vs base                │
_CodecDecodeLogs-10                                      3.649Mi ± 0%   3.649Mi ± 0%        ~ (p=0.364 n=10)
_CodecDecodeSamples-10                                   18.12Mi ± 0%   18.12Mi ± 0%        ~ (p=0.926 n=10)
_CodecDecodeSeries/application/vnd.google.protobuf-10    7.647Mi ± 0%   7.647Mi ± 0%        ~ (p=0.587 n=10)
_CodecDecodeSeries/application/json;_charset=utf-8-10    27.94Mi ± 0%   16.99Mi ± 0%  -39.18% (p=0.000 n=10)
_MergeResponses-10                                       2.362Mi ± 0%   2.408Mi ± 0%   +1.98% (p=0.000 n=10)
_UnwrapSeries-10                                        19.495Mi ± 0%   8.595Mi ± 0%  -55.91% (p=0.000 n=10)
_DecodeMergeEncodeCycle-10                               772.3Mi ± 0%   772.3Mi ± 0%        ~ (p=0.912 n=10)
geomean                                                  17.50Mi        14.54Mi       -16.91%

                                                      │ before.log  │              after.log               │
                                                      │  allocs/op  │  allocs/op   vs base                 │
_CodecDecodeLogs-10                                     41.10k ± 0%   41.10k ± 0%       ~ (p=1.000 n=10) ¹
_CodecDecodeSamples-10                                  411.9k ± 0%   411.9k ± 0%       ~ (p=1.000 n=10)
_CodecDecodeSeries/application/vnd.google.protobuf-10    32.00 ± 0%    32.00 ± 0%       ~ (p=1.000 n=10) ¹
_CodecDecodeSeries/application/json;_charset=utf-8-10   304.2k ± 0%   298.1k ± 0%  -2.01% (p=0.000 n=10)
_MergeResponses-10                                      100.1k ± 0%   100.1k ± 0%  -0.00% (p=0.002 n=10)
_UnwrapSeries-10                                        201.1k ± 0%   198.0k ± 0%  -1.54% (p=0.000 n=10)
_DecodeMergeEncodeCycle-10                              203.1k ± 0%   203.1k ± 0%       ~ (p=0.621 n=10)
geomean                                                 48.95k        48.70k       -0.51%
¹ all samples are equal
```

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [x] Tests updated
- [ ] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: pkg/querier/queryrange/codec.go [hunk 1]
```diff
@@ -866,7 +866,10 @@ func decodeResponseJSONFrom(buf []byte, req queryrangebase.Request, headers http
 		data := make([]logproto.SeriesIdentifier, 0, len(resp.Data))
 		for _, label := range resp.Data {
 			d := logproto.SeriesIdentifier{
-				Labels: label.Map(),
+				Labels: make([]*logproto.SeriesIdentifier_LabelsEntry, 0, len(label)),
+			}
+			for n, v := range label {
+				d.Labels = append(d.Labels, &logproto.SeriesIdentifier_LabelsEntry{Key: n, Value: v})
```
Comment (jeschkies) [inline pkg/querier/queryrange/codec.go:872]: I'd like to understand why this change is such a big improvement.

File: pkg/querier/queryrange/extensions.go [hunk 2]
```diff
@@ -43,6 +48,38 @@ func (m *LokiSeriesResponse) WithHeaders(h []queryrangebase.PrometheusResponseHe
 	return m
 }
 
+// UnmarshalJSON decodes from loghttpSeriesResponse JSON format directly into
+// the protobuf LokiSeriesResponse.
+func (m *LokiSeriesResponse) UnmarshalJSON(data []byte) error {
```
Comment (jeschkies) [inline pkg/querier/queryrange/extensions.go:53]: I need to test the error cases.

**Comment by jeschkies** [comment]
`MergeResponses-10` needs some investigation!

File: pkg/logproto/logproto.proto [hunk 3]
```diff
@@ -188,7 +188,11 @@ message SeriesResponse {
 }
 
 message SeriesIdentifier {
-  map<string, string> labels = 1;
+  message LabelsEntry {
+    string key = 1;
+    string value = 2;
+  }
```
Comment (chaudum) [inline pkg/logproto/logproto.proto:194]: Is there any advantage of making the message inline? I assume it should make it explicit that the type is only used in conjunction with SeriesIdentifier, right?

File: pkg/querier/queryrange/codec.go [hunk 4]
```diff
@@ -1194,7 +1183,7 @@ func (Codec) MergeResponse(responses ...queryrangebase.Response) (queryrangebase
 			for _, series := range lokiResult.Data {
 				// Use series hash as the key and reuse key
 				// buffer to avoid extra allocations.
-				key, keyBuffer = series.Hash(b, keyBuffer)
+				key = series.Hash(b)
```
Comment (chaudum) [inline pkg/querier/queryrange/codec.go:1186]: What's the reason for removing the `keyBuffer`? Or can you remove the comment above?

File: pkg/logproto/extensions.go [hunk 5]
```diff
@@ -16,39 +16,54 @@ import (
...
+		if entry.Key == key {
+			return entry.Value
+		}
+	}
+
+	return ""
+}
+
+func NewSeriesEntries(labels ...string) []SeriesIdentifier_LabelsEntry {
+	r := make([]SeriesIdentifier_LabelsEntry, 0, len(labels)/2)
```
Comment (chaudum) [inline pkg/logproto/extensions.go:62]: Maybe add a check if `len(labels)` is divisible by 2?

**Review by chaudum — COMMENTED** [review]
This looks great. I'm surprised by the benchmark numbers, tbh.

One thing I noticed is that `SeriesIdentifier_LabelsEntry` is the same data structure as `LabelAdapter`. Question is whether we can get rid of one.

**Comment by jeschkies** [comment]
> MergeResponses-10 needs some investigation!

Turns out that `sort.SliceFunc` is using temporary memory fro swapping elements.

File: pkg/logproto/logproto.proto (hunk 3 above)
Comment (jeschkies) [reply pkg/logproto/logproto.proto:194]: > I assume it should make it explicit that the type is only used in conjunction with SeriesIdentifier

Yes. That's the whole reason.

File: pkg/querier/queryrange/codec.go (hunk 4 above)
Comment (jeschkies) [reply pkg/querier/queryrange/codec.go:1185]: There's no need for a buffer anymore as we can sort the slice of labels directly.

File: pkg/logproto/extensions.go (hunk 5 above)
Comment (jeschkies) [reply pkg/logproto/extensions.go:62]: It's only used in tests but I will add it.

Comment (jeschkies) [reply pkg/logproto/extensions.go:65]: Added a panic following `labels.FromStrings`.

**Comment by jeschkies** [comment]
> I'm surprised by the benchmark numbers, tbh.

@chaudum what is surprising to you?

**Comment by chaudum** [comment]
> > I'm surprised by the benchmark numbers, tbh.
> 
> @chaudum what is surprising to you?

> geomean                                                         18.87m        12.51m        -33.70%

File: pkg/logproto/extensions.go (hunk 5 above)
Comment (MasslessParticle) [reply pkg/logproto/extensions.go:65]: If we're going to put a panic here, should we rename the function to something like `MustNewSeriesEntries`? This could help someone reading the code or if someone reaches for it outside of tests.

File: pkg/querier/querier_test.go [hunk 6]
```diff
@@ -237,9 +237,11 @@ func TestQuerier_SeriesAPI(t *testing.T) {
 	mockSeriesResponse := func(series []map[string]string) *logproto.SeriesResponse {
 		resp := &logproto.SeriesResponse{}
 		for _, s := range series {
-			resp.Series = append(resp.Series, logproto.SeriesIdentifier{
-				Labels: s,
-			})
+			id := logproto.SeriesIdentifier{}
```
Comment (MasslessParticle) [inline pkg/querier/querier_test.go:240]: We use this pattern to generate the Series in several places. Would it be worth pulling into a util?

File: pkg/logproto/extensions.go (hunk 5 above)
Comment (jeschkies) [reply pkg/logproto/extensions.go:65]: Yes. Good point.

**Comment by jeschkies** [comment]
> geomean 18.87m 12.51m -33.70%

That's because it's faster to iterate through a slice and the allocation is faster. Sometimes we just need to allocate one slice. Btw we save a lot of allocations by changing the slice into a slice of struct instead of a slice of pointers.

File: pkg/querier/querier_test.go (hunk 6 above)
Comment (jeschkies) [reply pkg/querier/querier_test.go:240]: Done
